presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageUtils.java

```diff
@@ -0,0 +1,939 @@
+        }
+        else if (vector instanceof Float8Vector) {
+            return buildBlockFromFloat8Vector((Float8Vector) vector, type);
+        }
+        else if (vector instanceof VarCharVector) {
+            if (type instanceof CharType) {
+                return buildCharTypeBlockFromVarcharVector((VarCharVector) vector, type);
+            }
+            else if (type instanceof TimeType) {
+                return buildTimeTypeBlockFromVarcharVector((VarCharVector) vector, type);
+            }
+            else {
+                return buildBlockFromVarCharVector((VarCharVector) vector, type);
+            }
+        }
+        else if (vector instanceof VarBinaryVector) {
+            return buildBlockFromVarBinaryVector((VarBinaryVector) vector, type);
+        }
+        else if (vector instanceof DateDayVector) {
+            return buildBlockFromDateDayVector((DateDayVector) vector, type);
+        }
+        else if (vector instanceof DateMilliVector) {
+            return buildBlockFromDateMilliVector((DateMilliVector) vector, type);
+        }
+        else if (vector instanceof TimeMilliVector) {
+            return buildBlockFromTimeMilliVector((TimeMilliVector) vector, type);
+        }
+        else if (vector instanceof TimeSecVector) {
+            return buildBlockFromTimeSecVector((TimeSecVector) vector, type);
+        }
+        else if (vector instanceof TimeStampSecVector) {
+            return buildBlockFromTimeStampSecVector((TimeStampSecVector) vector, type);
+        }
+        else if (vector instanceof TimeMicroVector) {
+            return buildBlockFromTimeMicroVector((TimeMicroVector) vector, type);
+        }
+        else if (vector instanceof TimeStampMilliTZVector) {
+            return buildBlockFromTimeMilliTZVector((TimeStampMilliTZVector) vector, type);
+        }
+        else if (vector instanceof ListVector) {
+            return buildBlockFromListVector((ListVector) vector, type);
+        }
+
+        throw new UnsupportedOperationException("Unsupported vector type: " + vector.getClass().getSimpleName());
+    }
+
+    public static Block buildBlockFromDictionaryVector(FieldVector fieldVector, FieldVector dictionaryVector)
+    {
+        // Validate inputs
+        requireNonNull(fieldVector, "encoded vector is null");
+        requireNonNull(dictionaryVector, "dictionary vector is null");
+
+        // Create a BlockBuilder for the decoded vector's data type
+        Type prestoType = getPrestoTypeFromArrowType(dictionaryVector.getField().getType());
+
+        Block dictionaryblock = null;
+        // Populate the block dynamically based on vector type
+        for (int i = 0; i < dictionaryVector.getValueCount(); i++) {
+            if (!dictionaryVector.isNull(i)) {
+                dictionaryblock = appendValueToBlock(dictionaryVector, prestoType);
+            }
+        }
+
+        // Get the Arrow indices vector
+        IntVector indicesVector = (IntVector) fieldVector;
```

Comment:
don't we need to check type before casting?

Reply:
done